crates/rune/src/modules/string.rs

```diff
@@ -1103,5 +1105,87 @@ fn parse_char(s: &str) -> Result<char, char::ParseCharError> {
     str::parse::<char>(s)
 }
 
+/// Returns the lowercase equivalent of this string slice, as a new [`String`].
+///
+/// 'Lowercase' is defined according to the terms of the Unicode Derived Core Property
+/// `Lowercase`.
+///
+/// Since some characters can expand into multiple characters when changing
+/// the case, this function returns a [`String`] instead of modifying the
+/// parameter in-place.
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```rune
+/// let s = "HELLO";
+///
+/// assert_eq!("hello", s.to_lowercase());
+/// ```
+///
+/// A tricky example, with sigma:
+///
+/// ```rune
+/// let sigma = "Σ";
+///
+/// assert_eq!("σ", sigma.to_lowercase());
+///
+/// // but at the end of a word, it's ς, not σ:
+/// let odysseus = "ὈΔΥΣΣΕΎΣ";
+///
+/// assert_eq!("ὀδυσσεύς", odysseus.to_lowercase());
+/// ```
+///
+/// Languages without case are not changed:
+///
+/// ```rune
+/// let new_year = "农历新年";
+///
+/// assert_eq!(new_year, new_year.to_lowercase());
+/// ```
+#[rune::function(instance)]
+fn to_lowercase(s: &str) -> VmResult<String> {
+    VmResult::Ok(vm_try!(String::try_from(s.to_lowercase())))
```

Comment:
This uses `to_lowercase` provided by `std`, which performs internal allocations using the default system allocator. In Rune we have to use types provided by [`rune-alloc`](https://docs.rs/rune-alloc) to ensure that memory limits are enforced.

In practice, I think this maybe means something like this:

```rust
let mut out = String::try_with_capacity(s.len());

for c in s.chars() {
    for c in c.to_uppercase() {
        output.push(c);
    }
}

VmResult::Ok(output)
```